File: src/rl_map_loader/map_manager.py
"""Map management and backup logic for RL Map Loader."""

import shutil
from pathlib import Path


class MapManager:
    """Handles map file operations, backups, and restoration."""

    TARGET_MAP = "Labs_Underpass_P.upk"
    BACKUP_EXT = ".bak"

    def __init__(self, rl_install_path: Path):
        self.rl_install_path = rl_install_path
        self.cooked_dir = self.rl_install_path / "TAGame" / "CookedPCConsole"
        self.target_file = self.cooked_dir / self.TARGET_MAP
        self.backup_file = self.cooked_dir / f"{self.TARGET_MAP}{self.BACKUP_EXT}"

    def is_valid_rl_install(self) -> bool:
        """Check if the path is a valid Rocket League installation."""
        return self.cooked_dir.exists() and self.cooked_dir.is_dir()

    def has_backup(self) -> bool:
        """Check if a backup of the original map exists."""
        # migration from backup file being .upk.upk.bak to .upk.bak
        old_backup = self.cooked_dir / f"{self.TARGET_MAP}.upk.bak"
        if old_backup.exists():
            try:
                old_backup.rename(self.backup_file)
            except Exception:
                pass

        return self.backup_file.exists()

    def has_custom_map_installed(self) -> bool:
        """Check if a custom map is currently installed (backup exists)."""
        return self.has_backup()

    def load_map(self, map_file: Path) -> tuple[bool, str]:
        """
        Load a custom map.

        If a backup doesn't exist, create one first.
        If a backup exists, don't create another one.

        Returns (success, message)
        """
        map_file = Path(map_file)

        if not map_file.exists():
            return False, f"Map file not found: {map_file}"

        if map_file.suffix.lower() != ".upk":
            return False, f"Invalid map file extension: {map_file.suffix}"

        if not self.cooked_dir.exists():
            self.cooked_dir.mkdir(parents=True, exist_ok=True)

        if not self.has_backup() and self.target_file.exists():
            try:
                shutil.copy2(self.target_file, self.backup_file)
            except Exception as e:
                return False, f"Failed to create backup: {e}"

        try:
            shutil.copy2(map_file, self.target_file)
            return True, f"Successfully loaded map: {map_file.name}"
        except Exception as e:
            return False, f"Failed to load map: {e}"

    def restore_original(self) -> tuple[bool, str]:
        """
        Restore the original map from backup.

        Removes the backup file after restoration.

        Returns (success, message)
        """
        if not self.has_backup():
            return False, "No backup found"

        try:
            shutil.copy2(self.backup_file, self.target_file)
            self.backup_file.unlink()
            return True, "Successfully restored original map"
        except Exception as e:
            return False, f"Failed to restore map: {e}"

File: src/map_manager.cpp
#include "map_manager.hpp"

#include <iostream>

namespace fs = std::filesystem;

MapManager::MapManager(const fs::path& rl_install_path)
    : rl_install_path(rl_install_path) {
    cooked_dir = this->rl_install_path / "TAGame" / "CookedPCConsole";
    target_file = cooked_dir / std::string(TARGET_MAP);
    backup_file = cooked_dir / (std::string(TARGET_MAP) + std::string(BACKUP_EXT));
    migrate_backup_format();
}

bool MapManager::is_valid_rl_install() const {
    return fs::exists(cooked_dir) && fs::is_directory(cooked_dir);
}

bool MapManager::has_backup() const {
    // Migration from backup file being .upk.upk.bak to .upk.bak
    fs::path old_backup = cooked_dir / (std::string(TARGET_MAP) + ".upk.bak");
    
    if (fs::exists(old_backup)) {
        try {
            fs::rename(old_backup, backup_file);
        } catch (const std::exception& e) {
            std::cerr << "Failed to migrate backup: " << e.what() << '\n';
        }
    }

    return fs::exists(backup_file);
}

bool MapManager::has_custom_map_installed() const {
    return has_backup();
}

std::pair<bool, std::string> MapManager::load_map(const fs::path& map_file) {
    auto map_file_path = fs::absolute(map_file);

    if (!fs::exists(map_file_path)) {
        return {false, "Map file not found: " + map_file_path.string()};
    }

    if (map_file_path.extension().string() != ".upk") {
        return {false, "Invalid map file extension: " + map_file_path.extension().string()};
    }

    if (!fs::exists(cooked_dir)) {
        try {
            fs::create_directories(cooked_dir);
        } catch (const std::exception& e) {
            return {false, std::string("Failed to create directory: ") + e.what()};
        }
    }

    if (!has_backup() && fs::exists(target_file)) {
        try {
            fs::copy_file(target_file, backup_file);
        } catch (const std::exception& e) {
            return {false, std::string("Failed to create backup: ") + e.what()};
        }
    }

    try {
        fs::copy_file(map_file_path, target_file, fs::copy_options::overwrite_existing);
        return {true, std::string("Successfully loaded map: ") + map_file_path.filename().string()};
    } catch (const std::exception& e) {
        return {false, std::string("Failed to load map: ") + e.what()};
    }
}

std::pair<bool, std::string> MapManager::restore_original() {
    if (!has_backup()) {
        return {false, "No backup found"};
    }

    try {
        fs::copy_file(backup_file, target_file, fs::copy_options::overwrite_existing);
        fs::remove(backup_file);
        return {true, "Successfully restored original map"};
    } catch (const std::exception& e) {
        return {false, std::string("Failed to restore map: ") + e.what()};
    }
}

void MapManager::migrate_backup_format() {
    // This is called automatically in constructor
    has_backup();
}
